xmppserver/src/main/java/org/jivesoftware/database/SequenceManager.java

```diff
@@ -77,6 +77,45 @@ public class SequenceManager {
     }
 
     private static final Cache<Integer, Data> sequenceBlocks = CacheFactory.createCache("Sequences");
+    private static final Map<String, Integer> uniqueNameToType = new ConcurrentHashMap<>();
+
+    /**
+     * Returns a SequenceManager instance for a given unique name with a default block size of 5.
+     * If a SequenceManager for the unique name already exists, it returns the existing instance.
+     * Otherwise, it creates a new SequenceManager with a unique type and the default block size.
+     *
+     * @param uniqueName the unique name for which to get the SequenceManager.
+     * @return the SequenceManager instance associated with the unique name.
+     */
+    public static SequenceManager getSequenceManagerByUniqueName(String uniqueName) {
+        return getSequenceManagerByUniqueName(uniqueName, 5);
+    }
+
+    /**
+     * Returns a SequenceManager instance for a given unique name. If a SequenceManager
+     * for the unique name already exists, it returns the existing instance. Otherwise,
+     * it creates a new SequenceManager with a unique type and the specified size.
+     *
+     * @param uniqueName the unique name for which to get the SequenceManager.
+     * @param size the block size for the SequenceManager if a new one is created.
+     * @return the SequenceManager instance associated with the unique name.
+     */
+    public static SequenceManager getSequenceManagerByUniqueName(String uniqueName, int size) {
+        // Return an existing SequenceManager if one exists for the unique name.
+        if (uniqueNameToType.containsKey(uniqueName)) {
+            return managers.get(uniqueNameToType.get(uniqueName));
+        }
+
+        // Otherwise, create a new SequenceManager for the unique name.
+        int type = 1;
+        while (true) {
+            if (!managers.containsKey(type)) {
+                uniqueNameToType.put(uniqueName, type);
+                return new SequenceManager(type, size);
+            }
+            type++;
+        }
+    }
```

Comment:
This basically replaces a unique number with a unique name, but that's still something that needs to be correctly correlated between implementations (granted, the risk of accidental duplicate value usage is a bit lower).

This is solving an issue that, to the best of my knowledge, hasn't existed in the past two decades. It adds a bit of complexity that we'll need to maintain, which I dislike.

I am, for example, wondering if this holds up in a cluster. The `uniqueNameToType` map doesn't seem to share its content across cluster nodes, which may be an issue.

Also, how does this behave after a restart, especially when various bits of code in Openfire get initialized in a potentially different order. Does this code guarantee that the same code will use the same sequence again? If that's not guaranteed, then things are likely to get broken (as these sequence values are typically used as identifiers for database content).

I'm not sure if the benefit of this change outweighs the added complexity. I prefer to stick with the previous approach, of simply using magic numbers. An alternative like the one proposed in this commit would at the very least require a fair bit of (automated) tests, before I'd be comfortable using this.

Reply:
Thanks Guus.
In the interest of time, no changes have now been made to SequenceManager.